clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+                                    GlobalVariable::InternalLinkage, CA,
+                                    "__sycl_native_cpu_decls");
+    auto *Begin = ConstantExpr::getGetElementPtr(GVar->getValueType(), GVar,
+                                                 getSizetConstPair(0, 0));
+    auto *End = ConstantExpr::getGetElementPtr(
+        GVar->getValueType(), GVar,
+        getSizetConstPair(0, NativeCPUEntries.size()));
+    return std::make_pair(Begin, End);
+  }
+
+  // Adds a global readonly variable that is initialized by given data to the
+  // module.
+  GlobalVariable *addGlobalArrayVariable(const Twine &Name,
+                                         ArrayRef<char> Initializer,
+                                         const Twine &Section = "") {
+    auto *Arr = ConstantDataArray::get(M.getContext(), Initializer);
+    auto *Var = new GlobalVariable(M, Arr->getType(), /*isConstant*/ true,
+                                   GlobalVariable::InternalLinkage, Arr, Name);
+    Var->setUnnamedAddr(GlobalValue::UnnamedAddr::Global);
+
+    SmallVector<char, 32> NameBuf;
+    auto SectionName = Section.toStringRef(NameBuf);
+    if (!SectionName.empty())
+      Var->setSection(SectionName);
+    return Var;
+  }
+
+  // Adds given buffer as a global variable into the module and returns a pair
+  // of pointers that point to the beginning and the end of the variable.
+  std::pair<Constant *, Constant *>
+  addArrayToModule(ArrayRef<char> Buf, const Twine &Name,
+                   const Twine &Section = "") {
+    auto *Var = addGlobalArrayVariable(Name, Buf, Section);
+    auto *ImageB = ConstantExpr::getGetElementPtr(Var->getValueType(), Var,
+                                                  getSizetConstPair(0, 0));
+    auto *ImageE = ConstantExpr::getGetElementPtr(
+        Var->getValueType(), Var, getSizetConstPair(0, Buf.size()));
+    return std::make_pair(ImageB, ImageE);
+  }
+
+  // Adds given data buffer as constant byte array and returns a constant
+  // pointer to it. The pointer type does not carry size information.
+  Constant *addRawDataToModule(ArrayRef<char> Data, const Twine &Name) {
+    auto *Var = addGlobalArrayVariable(Name, Data);
+    auto *DataPtr = ConstantExpr::getGetElementPtr(Var->getValueType(), Var,
+                                                   getSizetConstPair(0, 0));
+    return DataPtr;
+  }
+
+  // Creates all necessary data objects for the given image and returns a pair
+  // of pointers that point to the beginning and end of the global variable that
+  // contains the image data.
+  std::pair<Constant *, Constant *>
+  addDeviceImageToModule(ArrayRef<char> Buf, const Twine &Name,
+                         StringRef TargetTriple) {
+    // Create global variable for the image data.
+    // TODO: recheck TargetTriple.empty()
+    return addArrayToModule(
+        Buf, Name,
+        TargetTriple.empty() ? ""
+                             : "__CLANG_OFFLOAD_BUNDLE__sycl-" + TargetTriple);
```

Comment:
That long prefix causes inconveniences on Windows if I'm not mistaken, because there is a limit on section name.
Can we shorten it right away? If not, a `TODO` should be added to consider that later

Reply:
8 chars is pretty tight restriction if we are intended to have several targets at the same time.

Reply:
AFAIK, it does not cause any functional issues, but it affects your ability to inspect SYCL binaries, because you can't distinguish between sections and have to look into each other to understand what's in there.